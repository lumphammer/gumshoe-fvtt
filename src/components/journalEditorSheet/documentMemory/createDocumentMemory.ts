import { createStack } from "./createStack";
import { DocumentMemory } from "./types";

/**
 * Create a new empty document memory
 */
export function createDocumentMemory(
  period: number,
  maxDepth: number | null = null,
): DocumentMemory {
  return {
    stack: createStack(),
    serial: 0,
    period,
    maxDepth,
    state: "",
    snapshots: [""],
  };
}
