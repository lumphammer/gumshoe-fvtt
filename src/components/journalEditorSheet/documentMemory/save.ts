import createDiff from "textdiff-create";
import applyDiff from "textdiff-patch";

import { createStack } from "./createStack";
import { isMagicSerial } from "./isMagicSerial";
import { DocumentMemory, Edit, Stack } from "./types";

/**
 * This function pushes a new state onto a stack and returns the new stack.
 * Recurses down onto deeper stacks if necessary.
 */
function push(
  stack: Stack,
  newEdit: Edit,
  serial: number,
  period: number,
  depth: number,
  maxDepth: number | null,
  snapshots: string[],
): [Stack, string[]] {
  let nextStack = stack.next;

  // the new bombBay is the old bombBay plus the overflow from edits
  let bombBay: Edit[] = [...stack.bombBay];
  if (stack.edits.length >= period) {
    bombBay.push(stack.edits[0]);
  }

  // the new edits
  const edits = [
    ...stack.edits.slice(stack.edits.length === period ? 1 : 0),
    newEdit,
  ];

  let newSnapshots = snapshots;

  if (
    isMagicSerial(period, depth, serial) &&
    (maxDepth === null || depth < maxDepth)
  ) {
    // calculate the current snapshot by applying edits
    const [oldState = "", ...nextSnapshots] = snapshots;
    let newState = oldState;
    for (const edit of bombBay) {
      newState = applyDiff(newState, edit.changes);
    }
    const editToPush: Edit = {
      changes: createDiff(oldState, newState),
      timestamp: bombBay[bombBay.length - 1].timestamp,
      serial: bombBay[bombBay.length - 1].serial,
    };
    bombBay = [];

    let newNextSnapshots = [];
    [nextStack, newNextSnapshots] = push(
      nextStack || createStack(),
      editToPush,
      serial,
      period,
      depth + 1,
      maxDepth,
      nextSnapshots.length > 0 ? nextSnapshots : [""],
    );
    newSnapshots = [newState, ...newNextSnapshots];
  }

  return [
    {
      bombBay,
      edits,
      next: nextStack,
    },
    newSnapshots,
  ];
}

/**
 * Add a new state to the memory
 */
export function save(memory: DocumentMemory, state: string): DocumentMemory {
  if (state === memory.state) return memory;
  const serial = memory.serial + 1;
  const changes = createDiff(memory.state, state);
  const edit: Edit = {
    changes,
    timestamp: Math.floor(Date.now() / 1000),
    serial,
  };
  const [stack, snapshots] = push(
    memory.stack,
    edit,
    serial,
    memory.period,
    1,
    memory.maxDepth,
    memory.snapshots,
  );
  return {
    stack,
    serial,
    period: memory.period,
    maxDepth: memory.maxDepth,
    state,
    snapshots,
  };
}
