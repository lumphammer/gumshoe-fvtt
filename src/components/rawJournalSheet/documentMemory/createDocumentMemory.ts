import { createStack } from "./createStack";
import { DocumentMemory } from "./types";

/**
 * Create a new empty document memory
 */
export function createDocumentMemory(periodicity: number): DocumentMemory {
  return {
    stack: createStack(periodicity),
    serial: 0,
    period: periodicity,
  };
}
