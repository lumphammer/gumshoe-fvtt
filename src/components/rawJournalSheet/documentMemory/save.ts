import diff from "textdiff-create";

import { createStack } from "./createStack";
import { isMagicSerial } from "./isMagicSerial";
import { DocumentMemory, Edit, Stack } from "./types";

/**
 * This function pushes a new state onto a stack and returns the new stack.
 * Recurses down onto deeper stacks if necessary.
 */
function push(
  stack: Stack,
  newState: string,
  serial: number,
  period: number,
  depth: number,
): Stack {
  const newDelta: Edit = {
    change: diff(stack.snapshot, newState),
    timestamp: Math.floor(new Date().getTime() / 1000),
  };
  let next = stack.next;
  if (isMagicSerial(period, depth, serial)) {
    next = push(
      next ?? createStack(period),
      stack.snapshot,
      serial,
      period,
      depth + 1,
    );
  }
  return {
    edits: [...stack.edits.slice(1), newDelta],
    snapshot: newState,
    next,
  };
}

/**
 * Add a new state to the memory
 */
export function save(memory: DocumentMemory, state: string): DocumentMemory {
  const serial = memory.serial + 1;
  const stack = push(memory.stack, state, serial, memory.period, 1);
  return {
    stack,
    serial,
    period: memory.period,
  };
}
