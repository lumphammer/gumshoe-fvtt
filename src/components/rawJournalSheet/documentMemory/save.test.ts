import { afterAll, beforeEach, expect, test, vi } from "vitest";

import { createDocumentMemory } from "./createDocumentMemory";
import { save } from "./save";

const epoch = 0;

function advanceTime10s() {
  vi.setSystemTime(new Date().setSeconds(new Date().getSeconds() + 10));
}

beforeEach(() => {
  vi.setSystemTime(new Date(epoch));
});

afterAll(() => {
  vi.useRealTimers();
});

test("starts", () => {
  const h1 = createDocumentMemory(3);
  const h2 = save(h1, "foo");
  expect(h2).toEqual({
    stack: {
      edits: [
        null,
        null,
        {
          change: [[1, "foo"]],
          timestamp: expect.closeTo(epoch),
        },
      ],
      snapshot: "foo",
      next: null,
    },
    serial: 1,
    period: 3,
  });
});

test("stores a sequence of edits", () => {
  let h = createDocumentMemory(3);
  const states = ["foo", "foobar", "bar"];
  for (const state of states) {
    h = save(h, state);
    advanceTime10s();
  }
  expect(h).toMatchSnapshot();
});

test("stores a sequence of edits onto the next stack", () => {
  let h = createDocumentMemory(3);
  const states = ["foo", "foobar", "bar", "barbaz"];
  for (const state of states) {
    h = save(h, state);
    advanceTime10s();
  }
  expect(h).toMatchSnapshot();
});

test("stores a sequence of edits onto the third stack", () => {
  let h = createDocumentMemory(3);
  const states = [
    "a",
    "ab",
    "abc",
    "abcd",
    "abcde",
    "abcdef",
    "abcdefg",
    "abcdefgh",
    "abcdefghi",
    "abcdefghij",
  ];
  for (const state of states) {
    h = save(h, state);
    advanceTime10s();
  }
  expect(h).toMatchSnapshot();
});
