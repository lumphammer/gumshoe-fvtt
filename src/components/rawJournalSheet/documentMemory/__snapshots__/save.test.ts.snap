// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`stores a sequence of edits 1`] = `
{
  "maxDepth": null,
  "period": 3,
  "serial": 3,
  "snapshots": [
    "",
  ],
  "stack": {
    "bombBay": [],
    "edits": [
      {
        "changes": [
          [
            1,
            "a",
          ],
        ],
        "serial": 1,
        "timestamp": 0,
      },
      {
        "changes": [
          [
            0,
            1,
          ],
          [
            1,
            "b",
          ],
        ],
        "serial": 2,
        "timestamp": 10,
      },
      {
        "changes": [
          [
            0,
            2,
          ],
          [
            1,
            "c",
          ],
        ],
        "serial": 3,
        "timestamp": 20,
      },
    ],
    "next": null,
  },
  "state": "abc",
}
`;

exports[`stores a sequence of edits onto the next stack 1`] = `
{
  "maxDepth": null,
  "period": 3,
  "serial": 6,
  "snapshots": [
    "abc",
    "",
  ],
  "stack": {
    "bombBay": [],
    "edits": [
      {
        "changes": [
          [
            0,
            3,
          ],
          [
            1,
            "d",
          ],
        ],
        "serial": 4,
        "timestamp": 30,
      },
      {
        "changes": [
          [
            0,
            4,
          ],
          [
            1,
            "e",
          ],
        ],
        "serial": 5,
        "timestamp": 40,
      },
      {
        "changes": [
          [
            0,
            5,
          ],
          [
            1,
            "f",
          ],
        ],
        "serial": 6,
        "timestamp": 50,
      },
    ],
    "next": {
      "bombBay": [],
      "edits": [
        {
          "changes": [
            [
              1,
              "abc",
            ],
          ],
          "serial": 3,
          "timestamp": 20,
        },
      ],
      "next": null,
    },
  },
  "state": "abcdef",
}
`;

exports[`stores a sequence of edits onto the third stack 1`] = `
{
  "maxDepth": null,
  "period": 3,
  "serial": 21,
  "snapshots": [
    "abcdefghijklmnopqr",
    "abcdefghi",
    "",
  ],
  "stack": {
    "bombBay": [],
    "edits": [
      {
        "changes": [
          [
            0,
            18,
          ],
          [
            1,
            "s",
          ],
        ],
        "serial": 19,
        "timestamp": 180,
      },
      {
        "changes": [
          [
            0,
            19,
          ],
          [
            1,
            "t",
          ],
        ],
        "serial": 20,
        "timestamp": 190,
      },
      {
        "changes": [
          [
            0,
            20,
          ],
          [
            1,
            "u",
          ],
        ],
        "serial": 21,
        "timestamp": 200,
      },
    ],
    "next": {
      "bombBay": [],
      "edits": [
        {
          "changes": [
            [
              0,
              9,
            ],
            [
              1,
              "jkl",
            ],
          ],
          "serial": 12,
          "timestamp": 110,
        },
        {
          "changes": [
            [
              0,
              12,
            ],
            [
              1,
              "mno",
            ],
          ],
          "serial": 15,
          "timestamp": 140,
        },
        {
          "changes": [
            [
              0,
              15,
            ],
            [
              1,
              "pqr",
            ],
          ],
          "serial": 18,
          "timestamp": 170,
        },
      ],
      "next": {
        "bombBay": [],
        "edits": [
          {
            "changes": [
              [
                1,
                "abcdefghi",
              ],
            ],
            "serial": 9,
            "timestamp": 80,
          },
        ],
        "next": null,
      },
    },
  },
  "state": "abcdefghijklmnopqrstu",
}
`;
